veps/sig-compute/decentralized-migration/decentralized-migration-vep.md

```diff
@@ -0,0 +1,279 @@
+
+The actual migration channel is proxied through the `virt-handler` pods that are running on the source and target nodes. Any migration state information is passed through the `VirtualMachineInstance` migrationStatus. This is because `virt-handler` pods can only read/modify `VirtualMachineInstances`.
+
+The reason virt-launcher pods do not communicate directly with each other, is that it is not guaranteed that virt-launchers will have network connectivity. The virt-handler pods also have access to a migration network if one is defined.
+
+The migration channel in figure 1 is either the pod network, or a specially configured migration network.
+
+Figure 1 describes the current interaction between various components:
+![figure 1](migration_current_state.png)
+
+#### Proposed state
+To start a live migration one creates two `VirtualMachineInstanceMigration` resources. A `target` resource, and a `source` resource.
+
+The target cluster will have a synchronization controller running in the kubevirt namespace, that has access to the migration channel (either pod network, or migration network). This controller waits for incoming synchronization connections from a source. When a `source` migration resource is created in the cluster, this synchronization controller will connect to the target synchronization and start synchronizing the `VirtualMachineInstance` status between them.
+
+If the source and the target are in the same cluster, the connection will be inside of the synchronization controller.
+
+Once synchronization has been established, each `virt-handler` pod will interact with their respective `VirtualMachineInstance` the same way as before. The major difference being that it is possible there are two separate `VirtualMachineInstance`s.
+
+TBD: How to handle the source `VirtualMachine` once the migration has completed. We can shut down the source `VirtualMachine`, delete it, or keep it running. For now a good starting point is to simply shut down the `VirtualMachine` and someone else decide what to do with the `VirtualMachine`
+
+Figure 2 describes the interaction with a source and target `VirtualMachineInstanceMigration` in the same namespace
+
+![figure 2](decentralized_in_namespace.png)
+
+Figure 3 describes the interaction with a source and target `VirtualMachineInstanceMigration` in different namespaces.
+
+![figure 3](decentralized_cross_namespace.png)
+
+Figure 4 describes the interaction with a source and target `VirtualMachineInstanceMigration` in different clusters
+
+![figure 4](decentralized_cross_cluster.png)
+
+#### Synchronization controller
+The synchronization controller allows one to synchronize two distinct and separate `VirtualMachineInstance`s. The synchronization controller is a separate controller that has access to modify `VirtualMachineInstance`s and can connect to the migration network.
```

Comment:
It sounds like - but I'm missing the desctription  - as if the ctler is creating the target VMI.

But the target VMI depends on many things - NADs, DVs possibly, secrets, etc - who is creating those?
I assume somebody.

Now, why can "somebody" not also create the VMI?

If "somebody" (I guess the orechstrator) is responsible for creating - effectively - a working VM on the dst, including it's dependencies, then there is a clear cut.

Orch to create everything on the cluster (API), and libvirt to really own the migratoin.

Reply:
Because that somebody (orchestrator) is creating the VM and associated resources. And yes it should be a working VM, it is just stopped. If you don't migrate you should be able to start the VM (except that the disks are blank so it can't start). It can't start the VMI because it doesn't have enough information to properly start the VMI. In particular the node selector/node anti affinity/things of that nature. This is where the synchronization comes in. The synchronization makes it so the effective migration is not any different from what exists now. But the synchronization happens over a network connection instead of directly to the VMI

Reply:
Also VMI creation is normally the job of the kubevirt controller. Except when creating a stand alone VMI without a VM. However you cannot do this live migration without a VM resource on the target because of how storage live migration works. For the elusive use cases we will need to do a storage live migration.